libs/fn/src/lib/input/input-base.ts

```diff
@@ -0,0 +1,71 @@
+/**
+ * This abstract class contains common functionality that will be used by many different input components, such as the input, text area,
+ * select, multi-input etc.
+ */
+
+import { AfterViewInit, ChangeDetectorRef, Directive, Input, OnDestroy } from '@angular/core';
+import { DisabledBehavior, ReadonlyBehavior } from '@fundamental-ngx/fn/cdk';
+import { Subscription, merge } from 'rxjs';
+
+export type InputState = 'positive' | 'critical' | 'negative' | 'info';
+
+@Directive()
+export abstract class InputBase implements AfterViewInit, OnDestroy {
+    /** Placeholder for the input. */
+    @Input()
+    placeholder: string;
+
+    /** Whether or not this input is the 'display' type. */
+    @Input()
+    display: boolean;
+
+    /** The state of the input. */
+    @Input()
+    state: InputState;
+
+    /** Whether or not this input is disabled. */
+    disabled: boolean;
+
+    /** Whether or not this input is readonly. */
+    readonly: boolean;
+
+    /** @hidden */
+    private _subscriptions = new Subscription();
+
+    /** @hidden */
+    protected constructor(
+        protected _cdRef: ChangeDetectorRef,
+        protected disabled$?: DisabledBehavior,
+        protected readonly$?: ReadonlyBehavior
+    ) {}
+
+    /** @hidden */
+    ngAfterViewInit(): void {
+        this._setDisableReadonlyProperties();
+        this._listenToDisablingEvents();
+    }
+
+    /** @hidden */
+    ngOnDestroy(): void {
+        this._subscriptions.unsubscribe();
+    }
+
+    /** @hidden */
+    private _listenToDisablingEvents(): void {
```

Comment:
Now it's much better, but some same parts of code i found when reviewing @g-cheishvili's `feat/fn-tabs` branch. Maybe there is also some duplications with https://github.com/SAP/fundamental-ngx/pull/7790 I'm aware of getting into the same mess we currently have in core & platform.